### PR TITLE
add support for windows

### PR DESCRIPTION
**Pull request by aramase**
fixes https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/115

### CODE REVIEW

File: deploy/secrets-store-csi-driver-windows.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        beta.kubernetes.io/os: windows
+      serviceAccountName: secrets-store-csi-driver
+      containers:
+        - name: node-driver-registrar
+          image: mcr.microsoft.com/oss/kubernetes-csi/csi-node-driver-registrar:v1.3.0-pipeline
+          args:
+            - --v=5
+            - "--csi-address=unix://C:\\csi\\csi.sock"
+            - "--kubelet-registration-path=C:\\var\\lib\\kubelet\\plugins\\csi-secrets-store\\csi.sock"
+          env:
```
Comment (ritazh) [inline deploy/secrets-store-csi-driver-windows.yaml]: Is there an equivalent command we can run to remove the socket for preStop?
```
rm -rf /registration/secrets-store.csi.k8s.io-reg.sock
```

File: deploy/secrets-store-csi-driver-windows.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    metadata:
+      labels:
+        app: csi-secrets-store
+    spec:
+      nodeSelector:
+        beta.kubernetes.io/os: windows
+      serviceAccountName: secrets-store-csi-driver
+      containers:
+        - name: node-driver-registrar
+          image: mcr.microsoft.com/oss/kubernetes-csi/csi-node-driver-registrar:v1.3.0-pipeline
```
Comment (ritazh) [inline deploy/secrets-store-csi-driver-windows.yaml]: What is the equivalent of this image compare to `quay.io/k8scsi/csi-node-driver-registrar`? The latest at https://github.com/kubernetes-csi/node-driver-registrar/releases is `1.2.0`

Do we also need to update this tag for the linux deployment?

File: pkg/secrets-store/nodeserver.go [hunk 3]
```diff
@@ -282,22 +268,26 @@ func (ns *nodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
...
+			err = os.RemoveAll(file)
+			if err != nil {
+				log.Errorf("failed to remove file %s, err: %v", file, err)
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+		}
 	}
-	// Unmounting the target
-	err = mount.New("").Unmount(targetPath)
+	err := mount.CleanupMountPoint(targetPath, ns.mounter, false)
```
Comment (ritazh) [inline pkg/secrets-store/nodeserver.go]: Can you double check that `extensiveMountPointCheck` should be set to `false` for Windows, which uses `IsLikelyNotMountPoint`
https://github.com/kubernetes/utils/blob/master/mount/mount_helper_common.go#L26-L28

**Comment by ritazh** [comment]
Left few comments but mostly looks great! Would be great to think about how to add e2e tests for Windows as part of CI

File: Makefile [hunk 4]
```diff
@@ -39,8 +39,12 @@ sanity-test:
 	go test -v ./test/sanity
 build: setup
 	CGO_ENABLED=0 GOOS=linux go build -a -ldflags ${LDFLAGS} -o _output/secrets-store-csi ./cmd/secrets-store-csi-driver
+build-windows: setup
+	CGO_ENABLED=0 GOOS=windows go build -a -ldflags ${LDFLAGS} -o _output/secrets-store-csi.exe ./cmd/secrets-store-csi-driver
 image: build
 	docker build --no-cache -t $(IMAGE_TAG) -f Dockerfile .
+image-windows: build-windows
+	docker build --no-cache -t $(IMAGE_TAG) -f windows.Dockerfile .
```
Comment (ritazh) [inline Makefile]: nit: not sure what is the best naming convention for Dockerfile. here are few from other projects to consider:
https://github.com/moby/moby/blob/master/Dockerfile.windows
https://github.com/kubernetes/kubernetes/blob/master/test/images/busybox/Dockerfile_windows

WDYT?

File: pkg/secrets-store/nodeserver.go (hunk 3 above)
Comment (aramase) [reply pkg/secrets-store/nodeserver.go]: I've tested with `extensiveMountPointCheck` true too, but the List mountpoints interface is not implemented for windows which means it always returns not mount. I'll double check again.

File: deploy/secrets-store-csi-driver-windows.yaml (hunk 2 above)
Comment (aramase) [reply deploy/secrets-store-csi-driver-windows.yaml]: > What is the equivalent of this image compare to quay.io/k8scsi/csi-node-driver-registrar? The latest at https://github.com/kubernetes-csi/node-driver-registrar/releases is 1.2.0

There is still no official build for windows yet. We need the build from master as that contains all the changes required to make it work on windows. `v1.3.0-pipeline` is building the master.

> Do we also need to update this tag for the linux deployment?

Yes, we should update to the latest available release. I'll open a PR to bump it up.

File: deploy/secrets-store-csi-driver-windows.yaml (hunk 1 above)
Comment (aramase) [reply deploy/secrets-store-csi-driver-windows.yaml]: Will try it out and see if we can use a windows command to remove it

File: pkg/secrets-store/nodeserver.go (hunk 3 above)
Comment (aramase) [reply pkg/secrets-store/nodeserver.go]: Validated this! `IsMountPoint` done for extensive check calls list mount interface (which isn't implemented for windows). So it always returns not mnt.

File: deploy/secrets-store-csi-driver-windows.yaml (hunk 1 above)
Comment (aramase) [reply deploy/secrets-store-csi-driver-windows.yaml]: Added the prestop command to delete the socket

File: Makefile (hunk 4 above)
Comment (aramase) [reply Makefile]: Having the `*.Dockerfile` file naming convention makes it possible for editors to recognize this as a docker file and provides syntax highlighting. That was one of the reasons for naming it this way.

Comment (ritazh) [reply Makefile]: ok that makes sense!

**Comment by ritazh** [comment]
lgtm

**Pull request merged**